### PR TITLE
Port plugins to ts

### PR DESCRIPTION
**Pull request by marcusramberg**

### CODE REVIEW

**Comment by kraih** [comment]
Please rebase on main and run tests again.

File: src/plugins/ejs-engine.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
+import type {MojoContext} from '../types.js';
+import type {AsyncTemplateFunction} from 'ejs';
```
Comment (kraih) [inline src/plugins/ejs-engine.ts:2]: Not alphabetical. And maybe add a newline between type imports and other imports.

File: src/plugins/ejs-engine.ts [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+import {compile} from 'ejs';
+import type App from '../app.js';
+import {createHash} from 'crypto';
+import File from '../file.js';
+
+export default function ejsEnginePlugin (app: App): void {
+  app.renderer.addEngine('ejs', ejsEngine);
+}
+
+async function ejsEngine (ctx: MojoContext, options: {inline: string, viewPath: string}): Promise<Buffer> {
```
Comment (kraih) [inline src/plugins/ejs-engine.ts:12]: The options are wrong, we are going to need a `MojoEngineOptions` in `types.js` with all the possible ones as optional.

File: src/plugins/exception-helpers.ts [hunk 3]
```diff
@@ -11,32 +14,32 @@ export default function exceptionHelpersPlugin (app) {
...
+  const mode: string = ctx.app.mode;
+  if (await ctx.render({view: `not-found.${mode}`, maybe: true, status: 404}) === true) return true;
+  if (await ctx.render({view: 'not-found', maybe: true, status: 404}) === true) return true;
 
   const view = mode === 'development' ? 'mojo/debug' : 'mojo/not-found';
   return ctx.render({view, status: 404});
 }
 
-function httpException (ctx, error) {
+function httpException (ctx: MojoContext, error: Error): boolean {
```
Comment (kraih) [inline src/plugins/exception-helpers.ts:42]: Those are not `boolean`, rendering always results in a `Promise`.

File: src/plugins/json-config.ts [hunk 4]
```diff
@@ -0,0 +1,23 @@
+import type App from '../app.js';
+import File from '../file.js';
+
+export default function jsonConfigPlugin (app: App, options: {file: string}): void {
```
Comment (kraih) [inline src/plugins/json-config.ts:4]: `file` is optional.

File: src/plugins/view-helpers.ts [hunk 5]
```diff
@@ -0,0 +1,30 @@
+import type App from '../app.js';
+import type {InspectOptions} from 'util';
+import type {MojoContext, MojoStash} from '../types.js';
+import {inspect} from 'util';
+
+export default function viewHelpersPlugin (app: App): void {
+  app.decorateContext('inspect', (object: MojoStash, options: InspectOptions) => inspect(object, options));
+
+  app.addHelper('include',
+    (ctx: MojoContext, options: MojoStash, stash: MojoStash) => ctx.renderToString(options, stash));
```
Comment (kraih) [inline src/plugins/view-helpers.ts:10]: `options` should be a real `MojoRenderOptions` in `types.js` with all possible values as optional.

File: src/plugins/exception-helpers.ts (hunk 3 above)
Comment (marcusramberg) [reply src/plugins/exception-helpers.ts:42]: Doesn't httpException have to be async as well then?

File: src/plugins/ejs-engine.ts [hunk 6]
```diff
@@ -0,0 +1,36 @@
+import type {AsyncTemplateFunction} from 'ejs';
+import type {MojoContext, MojoEngineOptions} from '../types.js';
+
+import {compile} from 'ejs';
+import type App from '../app.js';
```
Comment (kraih) [inline src/plugins/ejs-engine.ts:5]: Not alphabetical.

File: src/types.ts [hunk 7]
```diff
@@ -26,6 +28,8 @@ export interface MojoStash {[key: string]: any}
 // eslint-disable-next-line @typescript-eslint/no-explicit-any
 export type MojoPlugin = (app: App, options: MojoStash) => any;
 
+export type MojoRenderOptions = ({maybe?: boolean, view?: string, status?: number} | string);
```
Comment (kraih) [inline src/types.ts:31]: What is the plain string? And where are `text`, `json` and so on?

Comment (marcusramberg) [reply src/types.ts:31]: plain string is https://github.com/mojolicious/mojo.js/blob/main/src/context/http.js#L26 - I guess that render method should actually take `MojoRenderOptions | string` instead.

File: src/plugins/exception-helpers.ts [hunk 8]
```diff
@@ -1,4 +1,8 @@
-export default function exceptionHelpersPlugin (app) {
+import type {MojoContext, MojoStash} from '../types.js';
+
```
Comment (kraih) [inline src/plugins/exception-helpers.ts:2]: Random newline.

Comment (kraih) [reply src/plugins/exception-helpers.ts:2]: Also order.

File: src/plugins/exception-helpers.ts [hunk 9]
```diff
@@ -11,32 +15,32 @@ export default function exceptionHelpersPlugin (app) {
...
 
-async function htmlException (ctx, error) {
-  ctx.stash.exception = error;
+async function htmlException (ctx: MojoContext, error: Error): Promise<boolean> {
+  (ctx.stash as MojoStash).exception = error;
 
-  const mode = ctx.app.mode;
-  if (await ctx.render({view: `exception.${mode}`, maybe: true, status: 500}) === true) return;
-  if (await ctx.render({view: 'exception', maybe: true, status: 500}) === true) return;
+  const mode: string = ctx.app.mode;
```
Comment (kraih) [inline src/plugins/exception-helpers.ts:26]: That seems odd, `app.mode` is already typed. Does it detect `ctx.app` correctly?

File: src/plugins/exception-helpers.ts [hunk 10]
```diff
@@ -11,32 +15,32 @@ export default function exceptionHelpersPlugin (app) {
...
-function exception (ctx, error) {
+function exception (ctx: MojoContext, error: Error): boolean {
   if (ctx.isWebSocket) return ctx.websocketException(error);
   return ctx.httpException(error);
 }
 
-async function htmlException (ctx, error) {
-  ctx.stash.exception = error;
+async function htmlException (ctx: MojoContext, error: Error): Promise<boolean> {
+  (ctx.stash as MojoStash).exception = error;
```
Comment (kraih) [inline src/plugins/exception-helpers.ts:24]: Same here, the type should be inferred from the context object. (Which is not ported yet, is it? Yea that's probably the reason...)

File: src/plugins/exception-helpers.ts [hunk 11]
```diff
@@ -63,27 +67,29 @@ function jsonException (ctx, error) {
...
-function txtNotFound (ctx) {
+async function txtNotFound (ctx: MojoContext): Promise<boolean> {
   return ctx.render({text: 'Not Found', status: 404});
 }
 
-function websocketException (ctx, error) {
+function websocketException (ctx: MojoContext, error: Error): void {
   ctx.log.error(error.stack);
+  // Can be removed once context has been ported
+  // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions
```
Comment (kraih) [inline src/plugins/exception-helpers.ts:93]: Looks like context really needs to be ported first.

File: src/types.ts (hunk 7 above)
Comment (kraih) [reply src/types.ts]: Agreed. 👍

File: src/plugins/ejs-engine.ts [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+
+    if (template === undefined) {
+      template = compile(options.inline, {async: true});
+      renderer.cache.set(checksum, template);
+    }
+  } else {
+    template = renderer.cache.get(options.viewPath);
+
+    if (options.viewPath === undefined) {
+      throw new Error('viewPath is not defined for ejsEngine!');
```
Comment (kraih) [inline src/plugins/ejs-engine.ts:28]: You don't have to shout! And i think that can be a one-liner.

Comment (kraih) [reply src/plugins/ejs-engine.ts:28]: Also, wouldn't this go before the cache lookup?

File: src/plugins/exception-helpers.ts [hunk 13]
```diff
@@ -14,32 +17,32 @@ export default function exceptionHelpersPlugin (app) {
   app.addHelper('websocketException', websocketException);
 }
 
-function exception (ctx, error) {
+function exception (ctx: MojoDualContext, error: Error): boolean {
```
Comment (kraih) [inline src/plugins/exception-helpers.ts:20]: This is wrong, `return ctx.httpException(error)` returns a promise.

Comment (kraih) [reply src/plugins/exception-helpers.ts:20]: Return values of these helpers should really be consistent.

File: src/plugins/exception-helpers.ts [hunk 14]
```diff
@@ -66,27 +69,27 @@ function jsonException (ctx, error) {
...
   return ctx.render({text: 'Internal Server Error', status: 500});
 }
 
-function txtNotFound (ctx) {
+async function txtNotFound (ctx: MojoDualContext): Promise<boolean> {
   return ctx.render({text: 'Not Found', status: 404});
 }
 
-function websocketException (ctx, error) {
+function websocketException (ctx: MojoDualContext, error: Error): void {
```
Comment (kraih) [inline src/plugins/exception-helpers.ts:92]: This one should probably also have a consistent return value.

File: src/plugins/exception-helpers.ts [hunk 15]
```diff
@@ -66,27 +69,28 @@ function jsonException (ctx, error) {
...
+async function txtNotFound (ctx: MojoDualContext): Promise<boolean> {
   return ctx.render({text: 'Not Found', status: 404});
 }
 
-function websocketException (ctx, error) {
+async function websocketException (ctx: MojoDualContext, error: Error): Promise<boolean> {
   ctx.log.error(error.stack);
-  if (ctx.isEstablished) ctx.ws.close(1011);
+  if ((ctx as WebSocketContext).isEstablished) ctx.ws.close(1011);
+  return await Promise.resolve(true);
```
Comment (kraih) [inline src/plugins/exception-helpers.ts:95]: That `await Promise.resolv()` does nothing.

File: src/plugins/exception-helpers.ts [hunk 16]
```diff
@@ -14,32 +17,32 @@ export default function exceptionHelpersPlugin (app) {
   app.addHelper('websocketException', websocketException);
 }
 
-function exception (ctx, error) {
+async function exception (ctx: MojoDualContext, error: Error): Promise<boolean> {
   if (ctx.isWebSocket) return ctx.websocketException(error);
-  return ctx.httpException(error);
+  return await httpException(ctx, error);
```
Comment (kraih) [inline src/plugins/exception-helpers.ts:22]: This no longer uses the helper. Break being able to change behaviour with new helpers.

**Pull request merged**